src/passes/OptimizeInstructions.cpp

```diff
@@ -1615,6 +1585,170 @@ struct OptimizeInstructions
     return nullptr;
   }
 
+  Expression* partialEvaluation(Binary* binary) {
+    assert(binary->right->is<Const>());
+    Const* right = binary->right->cast<Const>();
+
+    // detect overflow or zero during signed multiplication
+    auto hasSignedMulOverflow = [](auto a, auto b) {
+      using T = decltype(a);
+      auto min = std::numeric_limits<T>::min();
+      auto max = std::numeric_limits<T>::max();
+      return (a != T(0) && b != T(0) &&
+              ((b > T(0) && a > max / b) || (b < T(0) && a < max / b) ||
+               (b > T(0) && a < min / b) || (b < T(-1) && a > min / b)));
+    };
```

Comment:
It would be good to have a comment justifying why each of these cases is necessary and correct. Same for the unsigned versions below.